gimie/sources/gitlab.py

```diff
@@ -1,14 +1,287 @@
+
+        # unique to Gitlab Extractor
+        self.identifier = urlparse(data["id"]).path.split("/")[2]
+        self.sourceOrganization = self._get_organization(data["group"])
+
+        # Below are standard extracts
+        self.description = data["description"]
+        self.prog_langs = [lang["name"] for lang in data["languages"]]
+        self.date_created = datetime.fromisoformat(data["createdAt"][:-1])
+        self.date_modified = datetime.fromisoformat(
+            data["lastActivityAt"][:-1]
+        )
+        self.keywords = data["topics"]
+
+        # Get contributors as the project members that are not owners and those that have written merge requests
+        # owners are either multiple individuals or a group, if not user is marked as owner
+        user_author = [
+            user["node"]["user"]
+            for user in data["projectMembers"]["edges"]
+            if user["node"]["accessLevel"]["stringValue"] == "OWNER"
+        ]
+        author = user_author if len(user_author) > 0 else data["group"]
+        self.author = [self._get_author(author)]
+        # contributors are project members or merge request authors
+        project_members = [
+            user["node"]["user"]
+            for user in data["projectMembers"]["edges"]
+            if user["node"]["accessLevel"]["stringValue"] != "OWNER"
+        ]
+        merge_request_authors = [
+            author["node"]["author"]
+            for author in data["mergeRequests"]["edges"]
+        ]
+        duplicate_contributors = project_members + merge_request_authors
+        self.contributors = [
+            self._get_user(dict(unique))
+            for unique in set(
+                tuple(sorted(contrib.items()))
+                for contrib in duplicate_contributors
+            )
+        ]
+
+        if len(data["releases"]["edges"]) > 0:
+            # go into releases and take the name from the first node (most recent)
+            self.version = data["releases"]["edges"][0]["node"]["name"]
+            self.download_url = f"{self.path}/-/archive/{self.version}/{self.name}-{self.version}.tar.gz"
+
+        # for the license, we need to query the rest API
+        # the code below does not work, returns - if you have permission- the GitLab specific licence
+        # resp = requests.get(url=f"{GL_API_rest}/license/{self.identifier}")
+        # if resp.status_code == 200:
+        #     self.license = resp.json()
+
+    def _fetch_repo_data(self, url: str) -> Dict[str, Any]:
+        """Fetch repository metadata from GraphQL endpoint."""
+        response = query_repo_graphql(url, self._set_auth())
+        if "errors" in response:
+            raise ValueError(response["errors"])
+
+        return response["data"]["project"]
+
+    def _set_auth(self) -> Any:
+        """Set authentication headers for Gitlab API requests."""
+        try:
+            if not self.Gitlab_token:
+                self.Gitlab_token = os.environ.get("ACCESS_TOKEN")
```

Comment:
Now that we have _both_ GitLab and GitHub, maybe we should use more explicit variable names for the tokens. How about `GITHUB_TOKEN` and `GITLAB_TOKEN` ? If we change this, we would also have to modify the CI variables and `.env.dist`.

Reply:
I propose to leave this for the refactor PR